Routing config in PHP

willemverspyck opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Subject

<!-- Describe your Pull Request content here -->

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 4.x is for everything backwards compatible, like patches, features and deprecation notices
    - 5.x is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataAdminBundle/blob/4.x/CONTRIBUTING.md#base-branch
-->

I am targeting this branch, because it's BC.

Symfony is making XML routing configs deprecated, so I think it's a good to start to add the PHP config for routings in SonataAdminBundle. To prevent duplicate configs, I imported the PHP config from the XML config and made the XML config deprecated. The minimum required Symfony version for Sonata is 6.4, so PHP configs are fully supported.

Changes are based on https://github.com/symfony/symfony/pull/60232

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: https://keepachangelog.com/
    This will end up on https://github.com/sonata-project/SonataAdminBundle/releases,
    please keep it short and clear and to the point
-->

<!--
    If you are updating something that doesn't require
    a release, you can delete the whole "Changelog" section.
    (eg. update to docs, tests)
-->

<!-- REMOVE EMPTY SECTIONS
    - Don't repeat the verb in the messages (don't use "Added", "Changed", etc).
    - Don't use a full stop at the end of the messages.
-->
```markdown
### Added
- Added `sonata_admin.php` routing config

### Deprecated
- Importing `sonata_admin.xml` is deprecated
```

<!--
    If this is a work in progress, uncomment the "To do" section.
    You can add as many tasks as you want.
    If some are not relevant, just remove them.
-->
<!--
## To do

- [ ] Update the tests;
- [ ] Update the documentation;
- [ ] Add an upgrade note.
-->

willemverspyck committed:
Convert routing config to PHP

willemverspyck committed:
Add comment and strict_types

willemverspyck committed:
Import PHP config from the XML config

willemverspyck committed:
Trigger deprecation when XML config is used

willemverspyck committed:
Fix CI

dmaicher commented:
@willemverspyck thanks for starting this. Are you up for finalizing this? Symfony 8 release is aproaching and we will need to do this before we can add compatibility

willemverspyck committed:
Merge branch 'sonata-project:4.x' into 4.x

willemverspyck committed:
Use PHP config in tests

Fix Psalm

Fix CI

Update all config to PHP

Explicit set to PHP

Update config of tests to PHP

dmaicher commented on src/Resources/config/routing/sonata_admin.php:
```diff
@@ -0,0 +1,64 @@
+
+            if (false !== $realpath && __DIR__ === dirname($realpath)) {
+                @trigger_error(
+                    'The "sonata_admin.xml" routing configuration is deprecated since sonata-project/admin-bundle 4.83'
```
```suggestion
                    'The "sonata_admin.xml" routing configuration is deprecated since sonata-project/admin-bundle 4.39'
```

willemverspyck commented:
@dmaicher I got stuck back then in the tests with this error:
```
Symfony\Component\Config\Exception\LoaderLoadException: Cannot load resource "sonata_admin.php". Make sure there is a loader supporting the "php" type.
```

But just fixed this by also converting all routes in the tests to PHP configs.

Just did some tests. When I use the XML config in my application I get a deprecation notice:

```
User Deprecated: The "sonata_admin.xml" routing configuration is deprecated since sonata-project/admin-bundle 4.83 and will throw an error in 5.0. Import "sonata_admin.php" instead.
```

When I use the PHP config in my application, Sonata still works without deprecation notice! I just set the version in the deprecation notice to 4.83, for now.

dmaicher reviewed (commented):
Thanks a lot :+1: Tiny comment about the release version. It will be in the next minor which is `4.39`

dmaicher committed:
Update version for deprecation message

VincentLanglet commented:
Thanks
